packages/gator-permissions-snap/src/clients/accountApiClient.ts

```diff
@@ -64,129 +72,213 @@ export type AccountApiClientConfig = {
+   * @returns The token balance as BigInt.
    */
-  public async getTokenBalanceAndMetadata({
+  public async getTokenBalance({
     chainId,
     assetAddress,
     account,
+    retryOptions,
   }: {
     chainId: number;
     account: Hex;
     assetAddress?: Hex | undefined;
-  }): Promise<TokenBalanceAndMetadata> {
+    retryOptions?: RetryOptions;
+  }): Promise<bigint> {
     if (!chainId) {
       const message = 'No chainId provided to fetch token balance';
       logger.error(message);
-
       throw new InvalidInputError(message);
     }
 
     if (!account) {
       const message = 'No account address provided to fetch token balance';
       logger.error(message);
-
       throw new InvalidInputError(message);
     }
-
     const tokenAddress = assetAddress ?? AccountApiClient.#nativeTokenAddress;
 
-    const validatedResponse = await makeValidatedRequest(
-      `${this.#baseUrl}/tokens/${tokenAddress}?accountAddresses=${account}&chainId=${chainId}`,
-      {
-        timeoutMs: this.#timeoutMs,
-        maxResponseSizeBytes: this.#maxResponseSizeBytes,
-        fetch: this.#fetch,
-      },
-      TokenBalanceResponseSchema,
+    // Try up to initial attempt + retry attempts
+    const response = await this.#fetchTokenBalance(
+      tokenAddress,
+      account,
+      chainId,
+      retryOptions,
     );
 
-    const { accounts, type, iconUrl, symbol, decimals } = validatedResponse;
+    // Parse and validate the response
+    const { balances, count } = response;
 
-    const accountLowercase = account.toLowerCase();
-    const accountData = accounts.find(
-      (acc) => acc.accountAddress.toLowerCase() === accountLowercase,
-    );
+    // If count is 0 or balances array is empty, it means the token has no balance (0 balance)
+    if (count === 0 || balances.length === 0) {
+      logger.info(
+        `No balance found for token ${tokenAddress} on account ${account}`,
+      );
+      return BigInt(0);
+    }
 
-    if (!accountData) {
-      const message = `No balance data found for the account: ${account}`;
-      logger.error(message);
+    // Find the token in the balances array
+    const tokenData = balances.find(
+      (token) => token.address.toLowerCase() === tokenAddress.toLowerCase(),
+    );
 
-      throw new ResourceNotFoundError(message);
+    if (!tokenData) {
+      logger.error(`Token ${tokenAddress} not found in balance response`);
+      throw new ResourceNotFoundError(
+        `Token ${tokenAddress} not found in balance response`,
+      );
     }
 
-    if (
-      type !== undefined &&
-      !AccountApiClient.#supportedTokenTypes.includes(type)
-    ) {
-      const message = `Unsupported token type: ${type}`;
-      logger.error(message);
+    // Convert balance string to BigInt
+    // The balance comes as a decimal string in human-readable format (e.g., "1.5" for 1.5 ETH)
+    // We need to convert it to the smallest unit by multiplying by 10^decimals
```

Comment:
nit: Does the formatted string include all available dp?

I wonder if it could be problematic if the response is being rounded, and we are expecting an actual exact balance. However, I don't think we require the selected account to have a balance in the specified token, and only use this as a hint to the user.

Reply:
Thanks for the nit! The balance string from the API includes all available decimal places and is validated with a regex that allows arbitrary precision (/^\d+\.?\d*$/u on line 28). Since this is used as a user hint rather than for precise calculations, rounding shouldn’t cause any issues here.